fecgov/openFEC
# Fix incorrect counts when sort_hide_null=true

## Summary (required)

- Resolves #4626 

_Include a summary of proposed changes._
- adds function to determine whether count will be estimated or not (this avoids using `get_count` for the `is_estimate` return value which causes a count)
- use this function to:
  1. set `count=None` when the count is not going to be estimated--this lets the internal marshmallow pagination determine the count
  2. pass the estimated count along when `is_estimate=False`
- modifies signature of `join_committee_queries` so that total is not computed as an intermediate value


## How to test the changes locally
- two reviewers please.
- check out branch
- pytest
- please feel free to add additional relevant pytests for correct behavior
- performance tests are included below (the fix corrects the bug without adding performance overhead for the indicated test cases). The method was:
  1. day 0: determine a relevant subset of queries that would not be  estimates
  2. day 1: run queries on `dev`
  3. day 2: deploy PR branch to `dev` space. Rerun queries. This lag is necessary because caching preserves the original queries until an overnight refresh per @fec-jli 

```
# Schedule A
START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/schedules/schedule_a/?contributor_occupation=attorney&contributor_type=individual&api_key=DEMO_KEY&two_year_transaction_period=2020&recipient_committee_type=P&sort=-contribution_receipt_date&is_individual=true&sort_null_only=false&sort_hide_null=true&per_page=20" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":325649 at 247 seconds
# fix: "count":325649 at 248 seconds

START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/schedules/schedule_a/efile/?per_page=20&sort=-contribution_receipt_date&api_key=DEMO_KEY&sort_nulls_last=false&contributor_state=NJ&contributor_state=CA&sort_null_only=false&page=1&contributor_occupation=attorney&sort_hide_null=false" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":420037 at 125 seconds
# fix: "count":420542 at 127 seconds

# Candidates
START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/candidates/?api_key=DEMO_KEY&sort=name&page=1&sort_null_only=false&sort_nulls_last=false&sort_hide_null=false&per_page=20" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":43213 at 1 second
# fix: "count":43213 at 0 seconds

# Schedule B
START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/schedules/schedule_b/?min_amount=2000&per_page=20&sort=-disbursement_date&api_key=DEMO_KEY&max_amount=5000&sort_null_only=false&spender_committee_designation=J&spender_committee_designation=P&spender_committee_designation=U&spender_committee_org_type=C" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":309298 at 21 seconds
# fix: "count":309298 at 28 seconds

# Committees
START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/committees/?per_page=20&sort=name&api_key=DEMO_KEY&sort_nulls_last=false&sort_null_only=false&page=1&year=2016" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":14435 at 1 second
# fix: "count":14435 at 1 second

# Schedule E
START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/schedules/schedule_e/?per_page=20&sort_nulls_last=false&sort_hide_null=false&sort=-expenditure_date&api_key=DEMO_KEY&cycle=2020&sort_null_only=false" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":102198 at 12 seconds
# fix: "count":102380 at 9 seconds

START=$(date +%s)
curl -X GET "https://fec-dev-api.app.cloud.gov/v1/schedules/schedule_e/efile/?support_oppose_indicator=S&per_page=20&sort=-expenditure_date&api_key=DEMO_KEY&sort_nulls_last=false&sort_null_only=false&page=1&sort_hide_null=false" -H "accept: application/json"
END=$(date +%s)
DIFF=$(( $END - $START ))
echo "It took $DIFF seconds"
# "count":59299 at 5 seconds
# fix: "count":59463 at 5 seconds
```
